### PR TITLE
make session locks self renewal so clients don't have to implement th…

### PR DESCRIPTION
**Pull request by mpuncel**
…is behavior

### CODE REVIEW

File: pkg/kp/lock_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"testing"
+	"time"
+)
+
+const (
+	lockMessage = "Locked by lock_test.go"
+)
+
+func TestLock(t *testing.T) {
```
Comment (anthonybishopric) [inline pkg/kp/lock_test.go]: I want this test to be clearer about what it's verifying. From reading the test I can tell that it's establishing a lock and showing that a client that does not use the lock cannot access a key. However, I don't see tests that show that owning the lock does allow reads and writes.

Comment (anthonybishopric) [inline pkg/kp/lock_test.go]: While I don't want to have fully blown-out test names a la Rspec land, I do think it's valuable to know exactly what is being tested in the form of comments so we can establish how comprehensive coverage is at a glance.

Comment (mpuncel) [inline pkg/kp/lock_test.go]: I don't believe that locking does prevent reads or writes, we just use a convention that any process performing a replication must first acquire locks named a particular way.

File: pkg/kp/kv_test.go [hunk 2]
```diff
@@ -1,8 +1,6 @@
 package kp
 
-import (
-	"testing"
-)
+import "testing"
```
Comment (bwester) [inline pkg/kp/kv_test.go]: nit: Did an automatic tool make this change? I'd rather you keep the old formatting if possible.

File: pkg/kp/lock.go [hunk 3]
```diff
@@ -97,6 +145,11 @@ func (l Lock) Renew() error {
 
 // destroy a lock, releasing and deleting all the keys it holds
 func (l Lock) Destroy() error {
+	l.quitCh <- struct{}{}
```
Comment (bwester) [inline pkg/kp/lock.go]: `close(l.quitCh)` would be more idiomatic Go.

File: pkg/kp/lock.go [hunk 4]
```diff
@@ -86,6 +117,23 @@ func (l Lock) Lock(key string) error {
...
+		select {
+		case <-l.renewalCh:
+			err := l.Renew()
+			if err != nil {
+				l.renewalErrCh <- err
+				l.Destroy()
+				return
+			}
+		case <-l.quitCh:
+			close(l.renewalErrCh)
```
Comment (bwester) [inline pkg/kp/lock.go]: Could you boost this to the top of the method as `defer close(l.renewalErrCh)`?  That give the `Lock` object a nice invariant: the error channel will always be closed once its producer quits producing.

File: pkg/kp/kv_test.go (hunk 2 above)
Comment (mpuncel) [inline pkg/kp/kv_test.go]: not sure how that happened, i like the old format too.

File: pkg/kp/lock.go (hunk 4 above)
Comment (mpuncel) [inline pkg/kp/lock.go]: addressed

**Comment by anthonybishopric** [comment]
:+1:

**Comment by bwester** [comment]
:ship: Just make sure you squash your commits.

**Pull request merged**